lib/types.nix

```diff
@@ -160,118 +163,180 @@ let
+            merge ? mergeDefaultOption,
+            # Whether this type has a value representing nothingness. If it does,
+            # this should be a value of the form { value = <the nothing value>; }
+            # If it doesn't, this should be {}
+            # This may be used when a value is required for `mkIf false`. This allows the extra laziness in e.g. `lazyAttrsOf`.
+            emptyValue ? { },
+            # Return a flat attrset of sub-options. Used to generate
+            # documentation.
+            getSubOptions ? prefix: { },
+            # List of modules if any, or null if none.
+            getSubModules ? null,
+            # Function for building the same option type with a different list of
+            # modules.
+            substSubModules ? m: null,
+            # Function that merge type declarations.
+            # internal, takes a functor as argument and returns the merged type.
+            # returning null means the type is not mergeable
+            typeMerge ? defaultTypeMerge functor,
+            # The type functor.
+            # internal, representation of the type as an attribute set.
+            #   name: name of the type.
+            #   type: function to initialize the type with the passed values.
+            #   wrapped: the type wrapped in case of compound types.
+            #   payload: values passed to the function that initializes the type,
+            #            two payloads of the same type must be combinable with
+            #            the binOp binary operation.
+            #   binOp: binary operation that merge two payloads of the same type.
+            functor ? defaultFunctor self,
+            # The deprecation message to display when this type is used by an option
+            # If null, the type isn't deprecated
+            deprecationMessage ? null,
+            # The types that occur in the definition of this type. This is used to
+            # issue deprecation warnings recursively. Can also be used to reuse
+            # nested types
+            nestedTypes ? { },
+            __uncall__,
+            ...
+          }:
+          {
+            _type = "option-type";
+            inherit
+              name
+              check
+              merge
+              emptyValue
+              getSubOptions
+              getSubModules
+              substSubModules
+              typeMerge
+              deprecationMessage
+              nestedTypes
+              descriptionClass
+              __uncall__
+              ;
+            functor =
+              if functor ? wrappedDeprecationMessage then
+                functor
+                // {
+                  wrapped = functor.wrappedDeprecationMessage {
+                    loc = null;
+                  };
+                }
+              else
+                functor;
+            description = if description == null then self.name else description;
+          };
+
+        makeExtensibleOptionType =
```

Comment:
I have a bit of hard time reading this. I wonder if we could simplify this? For example using:

`lib.makeExtensible`

I don't fully understand the background behind the `recurseIntoArgs`. Why is this needed?
Can we use the existing tools provided by `lib.fixedPoints` this would probably reduce the local complexity